RM-7499: Add log messages for browser version and webdriver version

raffaelfoidl opened the pull request:
https://re-motion.atlassian.net/browse/RM-7499

Stefan-Ilic commented on Remotion/Web/Development.WebTesting/WebTestHelper.cs:
```diff
@@ -361,7 +407,7 @@ private DriverConfiguration MergeDriverConfiguration (DriverConfiguration config
           configurationOverride.RetryInterval ?? configuration.RetryInterval,
           configurationOverride.AsyncJavaScriptTimeout ?? configuration.AsyncJavaScriptTimeout);
     }
-	  
```
![image](https://user-images.githubusercontent.com/24362621/94439171-5822c180-01a0-11eb-8856-94f227114c4b.png)

Stefan-Ilic commented on Remotion/Web/Development.WebTesting/WebTestHelper.cs:
```diff
@@ -350,6 +353,49 @@ private void EnsureCursorIsOutsideBrowserWindow ()
+    private string GetWebdriverVersion ()
+    {
+      if (!(_mainBrowserSession.Driver.Native is IHasCapabilities driver))
+        return null;
```
Why did you choose to introduce null into the control flow instead of return "unknown" right here?

raffaelfoidl replied:
I was thinking about the very same issue while writing the code because I was not sure, but ended up using the `null`-way. The reasoning for this was in order to allow the caller of the respective method(s) to decide what to do with such a result using `null` to indicate this "special" situation.

Stefan-Ilic replied:
Alright, I see. I don't think it matters much in this case anyway, since the method is private and easily understandable to begin with 👍

MichaelKetting replied:
Fun fact. I just had a siilimar situation with MG. There it was the ProcessWrapper's Name, etc properties. I went with returning a string value all the time since a) there was no benefic to detecting if there's a version and b) it made the API non-null and thus removed complexity on the caller site.

So, here, I guess the same rules apply too. One way, we can just use the API to get a representative string value, the other way, we have to add extra code at the call site. And yes, if we wanted to add semantics to the information about there being a version, things would be very different, designwise.

raffaelfoidl replied:
I agree with your reasoning, especially due to the fact that  - as you say - there is no behavior tied to version information being present or not. I have updated the methods to make them return a string value instead of `null`.

Stefan-Ilic reviewed (changes requested):
Nice work, I'm just trying to understand some decision made

MichaelKetting reviewed (changes requested):
Is it possible to add this information also to our webtest exceptions?

raffaelfoidl commented:
> Is it possible to add this information also to our webtest exceptions?

Yes, in principle, it is. However, as far as I could tell, not for each and every instance.

**Possible, with further changes:**
* `Remotion.Web.Development.WebTesting.CoypuElementScopeClickAndFocusExtensions.Focus()` (L63), 2 of 3 calls adjusted easily
    - introduced a new optional `IDriver` parameter and adjusted call site

**Not evidently possible:**
1) `Remotion.Web.Development.WebTesting.CoypuElementScopeClickAndFocusExtensions.FocusClick` (L91), has 1 call
2) `Remotion.Web.Development.WebTesting.CoypuElementScopeExtensions.EnsureExistence()` (L58), has 28 calls
3) `Remotion.Web.Development.WebTesting.CoypuElementScopeExtensions.EnsureExistence()` (L62), has 28 calls
4) `Remotion.Web.Development.WebTesting.CoypuElementScopeExtensions.EnsureSingle()` (L89), has 4 calls
5) `Remotion.Web.Development.WebTesting.CoypuElementScopeExtensions.EnsureSingle()` (L93), has 4 calls
6) `Remotion.Web.Development.WebTesting.CoypuElementScopeExtensions.ExistsWorkaround()` (L158), has 55 calls
7) `Remotion.Web.Development.WebTesting.RequestErrorDetectionStrategies.CheckPageForError()` (L42), has  8 calls

**Overall:**
7 _direct_ calls could not evidently be altered as desired

**Reasons:**
1-7: only `ElementScope` object available in respective context - cannot retrieve `IDriver` from it

Stefan-Ilic commented on Remotion/Web/Development.WebTesting/Utilities/DriverExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+  /// <summary>
+  /// Extension methods for Coypu's <see cref="IDriver"/> interface.
+  /// </summary>
+  public static class DriverExtensions
```
I think we can unit test these methods, meaning we instance the drivers, put them in an `IDriver` variable and call these methods without needing to start a browser session, please verify if that's possible and if yes, put the tests in our "Web.Development.WebTesting.UnitTests" project (I hope that is its name, didn't check)

Stefan-Ilic commented on Remotion/Web/Development.WebTesting/WebTestException.cs:
```diff
@@ -25,6 +25,11 @@ public WebTestException (string message)
     }
 
+    public WebTestException (string message, string browser, string browserVersion, string driverVersion)
+        : base ($"{message}\n(Browser: {browser}, version {browserVersion})\n(Webdriver version: {driverVersion})")
```
Perhaps we should use `Environment.Newline` instead of `\n`, I think that's best practice

MichaelKetting replied:
actually, most of the time I just use `\r\n` because we're a) using Windows and thus always used Windows style and b) it's more readable than env.newline. Should we ever support Linux, we have to deal with this all over the place anyway. There are some projects with a more "universal" focus and a historical phase where we tried to use Env.Newline, but this just wasn't sustainable for the most part.

Stefan-Ilic reviewed (changes requested):
A few formatting changes, a request for unit tests and one decision I can't make on my own 😄

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocAutoCompleteReferenceValueControlObjectTest.cs:
```diff
@@ -461,7 +461,7 @@ public void TestSelectFirstMatch_NoFilterMatches_ThrowsWebTestException ()
           () => bocAutoComplete.SelectFirstMatch ("Invalid"),
           Throws.Exception.InstanceOf<WebTestException>()
-              .With.Message.EqualTo ("No matches were found for the specified filter: 'Invalid'."));
+              .With.Message.EqualTo (AssertionExceptionUtility.CreateExpectationException("No matches were found for the specified filter: 'Invalid'.", Driver).Message));
```
whitespace

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocCheckBoxControlObjectTest.cs:
```diff
@@ -64,7 +64,7 @@ public void TestIsDisabled_SetMethodsThrow ()
 
       Assert.That (control.IsDisabled(), Is.True);
-      Assert.That (() => control.SetTo (false), Throws.Exception.Message.EqualTo (AssertionExceptionUtility.CreateControlDisabledException ("SetTo").Message));
+      Assert.That (() => control.SetTo (false), Throws.Exception.Message.EqualTo (AssertionExceptionUtility.CreateControlDisabledException ("SetTo", Driver).Message));
```
I would not use the same reasoning or the concept of "mock" here, but the end result is the same. We care about a specific message and it's quite tedious to write this message by hand all the time and update the tests if the message changes. We do should one day consider if we don't want to make a custom Assertion that verifies the exception instance both on type and message, though. -> Ticket please

That's still not a test double, though.

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocListControlObjectTest.cs:
```diff
@@ -1170,7 +1171,8 @@ public void TestReadOnly_RowEditShouldThrow ()
           () => row.Edit(),
           Throws.InstanceOf<WebTestException>()
-              .With.Message.EqualTo ("The control is currently in a read-only state. Therefore, the operation is not possible."));
+              .With.Message.EqualTo (AssertionExceptionUtility.CreateExpectationException(
```
And I will stop the review until those are done.

iam-mholle commented on Remotion/Web/Development.WebTesting/Utilities/DriverExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+    /// <summary>
+    /// Gets the version of the browser associated with <paramref name="driver"/> or a <see cref="string"/> indicating that it could not be determined.
+    /// </summary>
+    public static string GetBrowserVersion (this IDriver driver)
```
I would love to see integration tests for `GetBrowserVersion` & `GetWebdriverVersion` (check if a version pattern is returned for the currently supported browser), just to be sure to update the capability keys if the browser manufacturers decided to change them (not unheard of).

iam-mholle reviewed (changes requested):
I don't really have something to add to the exception message changes.

iam-mholle commented on Remotion/Web/Development.WebTesting.IntegrationTests/DriverExtensionsTest.cs:
```diff
@@ -0,0 +1,47 @@
+namespace Remotion.Web.Development.WebTesting.IntegrationTests
+{
+  [TestFixture]
+  public class DriverExtensionsTest : IntegrationTest
```
I would also prefer an assertion making sure that a version pattern is returned, rather than "not `unknown`".

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocTreeViewControlObjectTest.cs:
```diff
@@ -539,7 +542,8 @@ public void TestNodeEnsureCollapsed ()
       var rootNode = bocTreeView.GetRootNode();
 
-      Assert.That (() => rootNode.Collapse(), Throws.InstanceOf<WebTestException>().With.Message.EqualTo ("TreeViewNode is already collapsed."));
+      Assert.That (() => rootNode.Collapse(), Throws.InstanceOf<WebTestException>().With.Message.EqualTo (
```
New line for Throws

MichaelKetting commented on Remotion/Web/Development.WebTesting/WebTestException.cs:
```diff
@@ -25,6 +25,11 @@ public WebTestException (string message)
     {
     }
 
+    public WebTestException (string message, string browser, string browserVersion, string driverVersion)
```
Since we don't use the ctors directly, we should move the driver-overload into the factory.

MichaelKetting commented on Remotion/Web/Development.WebTesting/Utilities/AssertionExceptionUtility.cs:
```diff
@@ -28,62 +29,69 @@ public static class AssertionExceptionUtility
     [MustUseReturnValue]
     [StringFormatMethod ("message")]
-    public static WebTestException CreateExpectationException ([NotNull] string message, params object[] args)
+    public static WebTestException CreateExpectationException ([NotNull] string message, IDriver driver = null, params object[] args)
```
format and args shouldn't be split. we need the driver to be the first, and then it can't be optional. since we always want a driver, that shouldn't be an issue, correct? this applies to all factory methods for consistency reasons. also are there downstream-dependencies?

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/DriverExtensionsTest.cs:
```diff
@@ -0,0 +1,50 @@
+    [Test]
+    public void GetBrowserName_ReturnsWebdriverVersion ()
+    {
+      Assert.That (ContainsVersionPattern (Driver.GetWebdriverVersion()), Is.True);
```
WebDriver's got an uppercase Driver

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/DriverExtensionsTest.cs:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    [Test]
+    public void GetBrowserName_ReturnsWebdriverVersion ()
```
Names

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/DriverExtensionsTest.cs:
```diff
@@ -0,0 +1,50 @@
+  [TestFixture]
+  public class DriverExtensionsTest : IntegrationTest
+  {
+    private static bool ContainsVersionPattern (string text)
```
Helpers at the bottom of the file.

MichaelKetting commented:
Also, please rebase when doing the changes.

MichaelKetting commented on Remotion/Web/Development.WebTesting.WebFormsControlObjects/TextBoxControlObject.cs:
```diff
@@ -55,10 +55,10 @@ public UnspecifiedPageObject FillWith (string text, IWebTestActionOptions action
 
       if (IsDisabled())
-        throw AssertionExceptionUtility.CreateControlDisabledException();
+        throw AssertionExceptionUtility.CreateControlDisabledException (driver: Driver);
```
redundant parameter name? remove everywhere.

MichaelKetting commented on Remotion/Web/Development.WebTesting/CoypuElementScopeExtensions.cs:
```diff
@@ -55,11 +55,11 @@ public static void EnsureExistence ([NotNull] this ElementScope scope)
       catch (MissingHtmlException exception)
       {
-        throw AssertionExceptionUtility.CreateControlMissingException (exception.Message);
+        throw AssertionExceptionUtility.CreateControlMissingException (null, exception.Message);
```
Add Driver parameter to method.

raffaelfoidl replied:
Unfortunately, I do not know how. I only have the `ElementScope` parameter in this method availabe, from which - from what I have gathered - I am not able to retrieve the desired driver.

iam-mholle replied:
We already needed/did this in `Remotion.Web.Development.WebTesting.CoypuElementScopeFillInWithAndSendKeysExtensions.GetDriver`. This API could be provided as an extension method. Not sure how to feel about this though.

```csharp
private static readonly Lazy<FieldInfo> s_driverFieldInfo = new Lazy<FieldInfo> (
    () => Assertion.IsNotNull (
        typeof (ElementScope).GetField ("_driver", BindingFlags.NonPublic | BindingFlags.Instance),
        "Coypu has changed, please update CoypuElementScopeFillInWithAndSendKeysExtensions.GetDriver() method."),
    LazyThreadSafetyMode.ExecutionAndPublication);

private static IDriver GetDriver ([NotNull] this ElementScope scope)
{
  ArgumentUtility.CheckNotNull ("scope", scope);

  return (IDriver) s_driverFieldInfo.Value.GetValue (scope);
}
```

@MichaelKetting

iam-mholle replied:
> Do you think we should?

Would definitely make things easier. 

> And ading a parameter or using the DriverScope.Driver, that's almost negligent if we didn't do this. Keep it internal for now, though.

What would the parameter solution be?

MichaelKetting replied:
requiring that the driver be passed as a parameter instead of getting it from scope.

MichaelKetting replied:
for the imediate task, we can use the reflection solution and then have a follow-up task for coypu. that way, we can clsoe this PR.

MichaelKetting commented on Remotion/Web/Development.WebTesting/CoypuElementScopeExtensions.cs:
```diff
@@ -86,11 +86,11 @@ public static void EnsureSingle ([NotNull] this ElementScope scope)
       }
       catch (MissingHtmlException exception)
       {
```
Add driver parameter to method.

raffaelfoidl replied:
see above

MichaelKetting commented on Remotion/Web/Development.WebTesting/CoypuElementScopeExtensions.cs:
```diff
@@ -155,7 +155,7 @@ public static bool ExistsWorkaround ([NotNull] this ElementScope scope)
       catch (AmbiguousException exception)
       {
-        throw AssertionExceptionUtility.CreateControlAmbiguousException (exception.Message);
+        throw AssertionExceptionUtility.CreateControlAmbiguousException (null, exception.Message);
```
add drivr parameter to method

raffaelfoidl replied:
see abiove

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/TreeViewControlObjectTest.cs:
```diff
@@ -298,17 +300,19 @@ public void TestSelectNodeWithIndexInHierarchy ()
       Assert.That (
           () => rootNode.GetNodeInHierarchy().WithIndex (999),
-          Throws.InstanceOf<WebTestException>().With.Message.EqualTo ("No node with the index '999' was found."));
+          Throws.InstanceOf<WebTestException>().With.Message.EqualTo (
```
formatting: linebreak before .With

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/IntegrationTest.cs:
```diff
@@ -44,6 +45,11 @@ protected WebTestHelper Helper
       get { return _webTestHelper; }
     }
 
+    protected IDriver Driver
```
[canbenull]

raffaelfoidl replied:
There are no JetBrains annotations in this specific assembly. Should a dependency be added or continued without the attribute?

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/DropDownMenuControlObjectTest.cs:
```diff
@@ -210,7 +210,8 @@ public void TestDropDownMenuControlObject_OpenDropDownMenuWithDelayGreaterThanTi
       var dropDownMenu = home.DropDownMenus().GetByLocalID ("MyDropDownMenu_DelayedLongerThanTimeout");
 
-      Assert.That (() => dropDownMenu.Open(), Throws.TypeOf<WebTestException>().With.Message.EqualTo ("Unable to open the menu."));
+      Assert.That (() => dropDownMenu.Open(), Throws.TypeOf<WebTestException>().With.Message.EqualTo (
```
linebreak and newlines

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/DriverExtensionsTest.cs:
```diff
@@ -0,0 +1,50 @@
+    [Test]
+    public void GetBrowserName_ReturnsWebDriverVersion ()
+    {
+      Assert.That (ContainsVersionPattern (Driver.GetWebDriverVersion()), Is.True);
```
Does.Match ()

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/DriverExtensionsTest.cs:
```diff
@@ -0,0 +1,50 @@
+    [Test]
+    public void GetBrowserName_ReturnsBrowserVersion ()
+    {
+      Assert.That (ContainsVersionPattern (Driver.GetBrowserVersion()), Is.True);
```
Does.Match ()

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocTreeViewControlObjectTest.cs:
```diff
@@ -698,7 +704,8 @@ public void TestContextMenuControlObject_OpenDropDownMenuWithDelayGreaterThanTim
       var contextMenu = node.GetContextMenu();
 
-      Assert.That (() => contextMenu.Open(), Throws.TypeOf<WebTestException>().With.Message.EqualTo ("Unable to open the menu."));
+      Assert.That (() => contextMenu.Open(), Throws.TypeOf<WebTestException>().With.Message.EqualTo (
```
format

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocListControlObjectTest.cs:
```diff
@@ -1565,7 +1572,8 @@ public void TestGoToPreviousPage_WithoutNavigator ()
       Assert.That (
           () => bocList.GoToPreviousPage(),
-          Throws.Exception.TypeOf<WebTestException>().With.Message.EqualTo ("Unable to change current page of the list. List only has one page."));
+          Throws.Exception.TypeOf<WebTestException>().With.Message.EqualTo (
```
format

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocListAsGridControlObjectTest.cs:
```diff
@@ -520,7 +521,8 @@ public void TestGoToSpecificPage ()
       Assert.That (
           () => bocList.GoToSpecificPage (1),
-          Throws.Exception.TypeOf<WebTestException>().With.Message.EqualTo ("Unable to change current page of the list. List is currently in edit mode."));
+          Throws.Exception.TypeOf<WebTestException>().With.Message.EqualTo (
```
foramt

MichaelKetting commented on Remotion/Web/Development.WebTesting/Utilities/AssertionExceptionUtility.cs:
```diff
@@ -28,62 +29,70 @@ public static class AssertionExceptionUtility
     [NotNull]
     [MustUseReturnValue]
-    public static WebTestException CreateControlDisabledException ([CallerMemberName] string operationName = "")
+    public static WebTestException CreateControlDisabledException (IDriver driver, [CallerMemberName] string operationName = "")
```
When all is done, I'm thinking we may be able to do a not-null on trunk and have an overload without the driver for the backport.

MichaelKetting commented on Remotion/Web/Development.WebTesting/Utilities/AssertionExceptionUtility.cs:
```diff
@@ -28,62 +29,70 @@ public static class AssertionExceptionUtility
   {
```
I believe my remarks regarding making the driver parameter not-nullable (and then only adding an overload in the backport got lost.

MichaelKetting commented on Remotion/Web/Development.WebTesting/CoypuElementScopeClickAndFocusExtensions.cs:
```diff
@@ -56,8 +56,11 @@ public static void ContextClick ([NotNull] this ElementScope scope, [NotNull] We
+    /// </param>
     /// <exception cref="WebTestException">The element is currently disabled.</exception>
-    public static void Focus ([NotNull] this ElementScope scope)
+    public static void Focus ([NotNull] this ElementScope scope, IDriver driver = null)
```
get from scope

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/TreeViewControlObjectTest.cs:
```diff
@@ -95,7 +96,8 @@ public void TestNodeIsChecked ()
           () => rootNode.IsChecked(),
           Throws.InstanceOf<WebTestException>()
-              .With.Message.EqualTo ("The checkbox could not be found: Unable to find xpath: ./tbody/tr/td[a[contains(@onclick, 'TreeView_SelectNode')]]/input[@type='checkbox']"));
+              .With.Message.EqualTo (
```
I'm missing the consistency in the formatting. this would be the preferred style for lines we have reformatted in this PR and the message does not fit in the With.Message.EqualTo line.

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/DropDownMenuControlObjectTest.cs:
```diff
@@ -59,23 +59,27 @@ public void TestIsDisabled_SetMethodsThrow ()
-      Assert.That (() => disabledControl.GetItemDefinitions(), Throws.Exception.Message.EqualTo (AssertionExceptionUtility.CreateControlDisabledException ("GetItemDefinitions").Message));
+      Assert.That (
+          () => disabledControl.GetItemDefinitions(),
+          Throws.Exception.Message.EqualTo (AssertionExceptionUtility.CreateControlDisabledException (Driver, "GetItemDefinitions").Message));
```
Since we're already reformatting, please add "With" before "Message"

MichaelKetting commented on Remotion/Web/Development.WebTesting.ControlObjects/DropDownMenuControlObjectBase.cs:
```diff
@@ -128,12 +128,12 @@ UnspecifiedPageObject IFluentControlObjectWithSelectableItems.WithItemID (string
 
       if (itemCommand.IsDisabled())
-        throw AssertionExceptionUtility.CreateCommandDisabledException (operationName: operationName);
+        throw AssertionExceptionUtility.CreateCommandDisabledException (Driver, operationName);
```
why remove the name here?

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocTreeViewControlObjectTest.cs:
```diff
@@ -506,10 +510,12 @@ public void TestEvaluatedNodeExpandable ()
       Assert.That (
           () => node.Expand(),
-          Throws.InstanceOf<WebTestException>().With.Message.EqualTo ("The WebTreeViewNode cannot be expanded as it has no children."));
+          Throws.InstanceOf<WebTestException>().With.Message.EqualTo (
```
reformat

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocListControlObjectTest.cs:
```diff
@@ -1667,7 +1681,8 @@ public void TestGoToLastPage_WithoutNavigator ()
       Assert.That (
           () => bocList.GoToLastPage(),
-          Throws.Exception.TypeOf<WebTestException>().With.Message.EqualTo ("Unable to change current page of the list. List only has one page."));
+          Throws.Exception.TypeOf<WebTestException>().With.Message.EqualTo (
```
reformat

MichaelKetting commented on Remotion/ObjectBinding/Web.Development.WebTesting.IntegrationTests/BocListAsGridControlObjectTest.cs:
```diff
@@ -532,7 +535,8 @@ public void TestGoToFirstPage ()
       Assert.That (
           () => bocList.GoToFirstPage(),
-          Throws.Exception.TypeOf<WebTestException>().With.Message.EqualTo ("Unable to change current page of the list. List is currently in edit mode."));
+          Throws.Exception.TypeOf<WebTestException>().With.Message.EqualTo (
```
reformat

MichaelKetting commented on Remotion/Web/Development.WebTesting/CoypuElementScopeExtensions.cs:
```diff
@@ -31,6 +33,12 @@ namespace Remotion.Web.Development.WebTesting
+    private static readonly Lazy<FieldInfo> s_driverFieldInfo = new Lazy<FieldInfo> (
+        () => Assertion.IsNotNull (
+            typeof (ElementScope).GetField ("_driver", BindingFlags.NonPublic | BindingFlags.Instance),
+            "Coypu has changed, please update CoypuElementScopeExtensions.GetDriver() method."),
```
Please drop the 'please' from the exception message.

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/TextBoxControlObjectTest.cs:
```diff
@@ -87,10 +89,10 @@ public void TestIsReadOnly_SetMethodsThrow ()
       Assert.That (
           () => control.FillWith ("text", FinishInput.Promptly),
-          Throws.Exception.Message.EqualTo (AssertionExceptionUtility.CreateControlReadOnlyException().Message));
+          Throws.Exception.Message.EqualTo (AssertionExceptionUtility.CreateControlReadOnlyException (Driver).Message));
```
With

MichaelKetting commented on Remotion/Web/Development.WebTesting.IntegrationTests/ElementScopeTest.cs:
```diff
@@ -37,7 +37,7 @@ public void TestFocus_IsDisabled ()
       Assert.That (
           () => disabledButton.Focus(),
-          Throws.Exception.Message.EqualTo (AssertionExceptionUtility.CreateControlDisabledException ("Focus").Message));
+          Throws.Exception.Message.EqualTo (AssertionExceptionUtility.CreateControlDisabledException (Driver, "Focus").Message));
```
With

raffaelfoidl commented:
Added to master with 736a8e227d2501ae6bcfab8bee276b53c311cb01.
